### PR TITLE
Initial recipe editor implementation

### PR DESCRIPTION
**Pull request by ayan4m1**
* [Rename](https://github.com/mixnjuice/frontend/pull/58/files#diff-be5bcbffb67f29d9a5b5d57f8e1e6eb3R53) Calculator to RecipeEditor
* [Add](https://github.com/mixnjuice/frontend/pull/58/files#diff-08fbd214821717845b939de9ad39aa0e) FlavorBrowser component
* [Add](https://github.com/mixnjuice/frontend/pull/58/files#diff-30db77d7ab6ff0242c46162f2c90e480) IngredientList component
* [Add](https://github.com/mixnjuice/frontend/pull/58/files#diff-412987e45f8c2050571d0a7e9791612a) RecipeComponents component
* [Add](https://github.com/mixnjuice/frontend/pull/58/files#diff-c0e1bed450c41b8555bf0c101b86e39b) RecipeDetails component
* Fix issue where [Login](https://github.com/mixnjuice/frontend/pull/58/files#diff-0d5d449f3d04b75da238685ca6505e9dR43) would not redirect correctly if you opened a bookmarked login link
* [Replace](https://github.com/mixnjuice/frontend/pull/58/files#diff-98b8603c6bace53a25539a068110f30aL257) Recipe page implementation with RecipeDetails component usage
* Partially implement Flavors page
* Partially implement FlavorStash page
* Implement RecipeEditor page
* [Fix](https://github.com/mixnjuice/frontend/pull/58/files#diff-4f4551829dd78334f41b7e816d80b553L54) issue where container width would not expand past 1200px on HiDPI monitors
* [Refactor](https://github.com/mixnjuice/frontend/pull/58/files#diff-4f4551829dd78334f41b7e816d80b553R285) "change disabled buttons to gray" use SCSS nesting
* [Add](https://github.com/mixnjuice/frontend/pull/58/files#diff-4f4551829dd78334f41b7e816d80b553R298) "sm" button modifier

Things currently missing from the RecipeEditor implementation:

* Robust form validation
* Redirect to the recipe URL after save (needs backend work)

### CODE REVIEW

File: src/components/RecipeComponents/RecipeComponents.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import { Table, OverlayTrigger, Tooltip } from 'react-bootstrap';
+
+export default class RecipeComponents extends Component {
+  static propTypes = {
+    components: PropTypes.arrayOf(
+      PropTypes.shape({
+        name: PropTypes.string,
+        density: PropTypes.number,
+        percentage: PropTypes.number,
+        mililiters: PropTypes.number,
```
Comment (daviddyess) [inline src/components/RecipeComponents/RecipeComponents.js]: `milliliters` ?

File: src/pages/Login.js [hunk 2]
```diff
@@ -40,7 +40,7 @@ export class Login extends Component {
     const { loggingIn, loggedIn, lastLocation } = this.props;
 
     if (loggedIn) {
-      return <Redirect to={lastLocation.pathname} />;
+      return <Redirect to={lastLocation ? lastLocation.pathname : '/'} />;
```
Comment (daviddyess) [inline src/pages/Login.js]: :+1:

File: src/pages/RecipeEditor.js [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+          </Modal.Header>
+          <Modal.Body>
+            <p>If you delete this recipe, it will be gone forever and ever.</p>
+          </Modal.Body>
+          <Modal.Footer>
+            <Button variant="secondary" onClick={this.handleDeleteDismiss}>
+              Close
+            </Button>
+            <Button variant="primary" onClick={this.handleDeleteConfirm}>
+              Save changes
```
Comment (daviddyess) [inline src/pages/RecipeEditor.js]: Maybe change this to `Confirm Deletion` ?

File: src/pages/RecipeEditor.js [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+                          name="desiredNicotineStrength"
+                          type="number"
+                          value={desiredNicotineStrength}
+                          onChange={this.handleUserInput}
+                          placeholder="0"
+                          required
+                        />
+                        <InputGroup.Append>
+                          <InputGroup.Text id="amount-unit">
+                            mg/mL
```
Comment (daviddyess) [inline src/pages/RecipeEditor.js]: I've seen `ml` other places, should keep it consistent. I tend to prefer mL, for what it's worth.

File: src/utils/index.js [hunk 5]
```diff
@@ -24,15 +24,22 @@ export function buildActions(reducer, actions) {
...
 export const mockComponent = (name, props = {}) => () =>
   createElement(name, props, props.children);
 
+/**
+ * Creates a mock component which will expose its props for snapshot testing purposes
+ * @param {*} name The component/element name (in PascalCase)
+ */
+export const mockComponentWithProps = name => props =>
+  createElement(name, props, props.children);
+
```
Comment (daviddyess) [inline src/utils/index.js]: :100:

File: src/pages/RecipeEditor.js (hunk 4 above)
Comment (ayan4m1) [reply src/pages/RecipeEditor.js]: OK, I think I got them all.

**Pull request merged**